sfdx-source/apex-common/main/classes/fflib_Criteria.cls

```diff
@@ -0,0 +1,654 @@
+	{
+		return notInSet(field, new Objects(new List<Double>(values)));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Set<Id> values)
+	{
+		return notInSet(field, new Objects(new List<Id>(values)));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Set<Integer> values)
+	{
+		return notInSet(field, new Objects(new List<Integer>(values)));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Set<Long> values)
+	{
+		return notInSet(field, new Objects(new List<Long>(values)));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Set<String> values)
+	{
+		return notInSet(field, new Objects(new List<String>(values)));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Set<Object> values)
+	{
+		return notInSet(field, new Objects(values));
+	}
+	public fflib_Criteria notInSet(Schema.SObjectField field, Objects values)
+	{
+		evaluators.add(new FieldSetEvaluator(field, fflib_Operator.NOT_IN, values));
+		return this;
+	}
+
+	/**
+	 * Generates the SOQL equivalent of the criteria provided
+	 *
+	 * @return The "where" part in the SOQL statement
+	 */
+	public String toSOQL()
+	{
+		if (this.evaluators.isEmpty()) return '';
+
+		String result = '';
+		for (Evaluator evaluator : this.evaluators)
+		{
+			if (result != '')
+			{
+				result += ' ' + this.comparatorType + ' ';
+			}
+			result += evaluator.toSOQL();
+		}
+
+		if (this.embraced) result = '(' + result + ')';
+
+		return result;
+	}
+
+	private static String operatorToString(fflib_Operator operator)
+	{
+		if (operator == fflib_Operator.EQUALS)
+			return '=';
+		else if (operator == fflib_Operator.NOT_EQUALS)
+			return '!=';
+		else if (operator == fflib_Operator.LESS_THAN)
+			return '<';
+		else if (operator == fflib_Operator.LESS_THAN_OR_EQUAL_TO)
+			return '<=';
+		else if (operator == fflib_Operator.GREATER_THAN)
+			return '>';
+		else if (operator == fflib_Operator.GREATER_THAN_OR_EQUAL_TO)
+			return '>=';
+		else if (operator == fflib_Operator.LIKEx)
+			return ' like ';
+		else if (operator == fflib_Operator.INx)
+			return ' IN ';
+		else if (operator == fflib_Operator.NOT_IN)
+			return ' NOT IN ';
+		else if (operator == fflib_Operator.INCLUDES)
+			return ' INCLUDES ';
+		else if (operator == fflib_Operator.EXCLUDES)
+			return ' EXCLUDES ';
+		else
+				return null;
+	}
+
+	/**
+	 * Adds braces to the condition when generated to SOQL
+	 *
+	 * @param embraced Braces will be added if set to TRUE
+	 */
+	private void setEmbraced(Boolean embraced)
+	{
+		this.embraced = embraced;
+	}
+
+	/**
+	 * @param value The value to convert
+	 *
+	 * @return Returns the given value converted to literal string
+	 */
+	private static String toLiteral(final Object value)
+	{
+		if (value == null) return 'null';
+
+		if (value instanceof String || value instanceof Id)
+		{
+			String manipulated = (String) value;
+			return '\'' + manipulated + '\'';
```

Comment:
Can the class have a toggle for adding “String.EscapeSingleQuotes()” for the string data type? For example,
If (SOQLInjectionProtection == true) manipulated = String.EscapeSingleQuotes(manipulated);

Reply:
I understand what you mean, but I do not think that belongs to the purpose of this class. I am a huge fan of the SOLID design principles. The purpose of this class is to control/hold a criteria, not to go too deep in all the posibilites of a SOQL statement. Maybe I should try to see if I can split this behaviour into a separate class, as I even think that generating the criteria to a SOQL is already pushing the "limits" of the scope of this class.

Reply:
A separate implementation class might be the way to go, although what is here is so very close to the need.  I think offering SOQL criteria creation without injection protection is risky, especially where the rest of the library delivers CRUD and FLS protection (related to DML and SELECT statements). Not having injection protection could discourage adoption, or encourage local customization that makes upgrading more difficult in the future. Maybe the amount of change can be minimized by putting the specifics of SOQL (and its injection management) in a separate “Evaluator” implementation.

Reply:
My main concern is that this isn't a "SOQL criteria creation" feature in the first place, it is a generic criteria feature and one of the places where it can be used is for rendering SOQL condition strings. 

The main benefit of this feature it to make complex `IF` conditions more readable:
```
    public Accounts getByLargeAccountsInCountry(String countryName)
    {
        return new AccountsImp(
                getRecords(
                       new AccountCriteria()
                               .ShippingCountryEquals(countryName)
                               .NumberOfEmployeesGreaterThan(numberOfEmployees)
                               . ......
                               . ....
                               . ...
                )
        );
    }
```
In addition to this, as I saw that many of these conditions in Domain and Selectors are the same, you can also generate it to SOQL.
I will look into a separate SOQL generator class, something like: 
`String queryWhereString = fflib_CriteriaSOQL.toSOQL(criteria);`